drivers/regulator/regulator_npm1300.c

```diff
@@ -0,0 +1,593 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nordic_npm1300_regulator
+
+#include <errno.h>
+#include <string.h>
+
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/regulator.h>
+#include <zephyr/dt-bindings/regulator/npm1300.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/linear_range.h>
+#include <zephyr/sys/util.h>
+
+/* nPM1300 voltage sources */
+enum npm1300_sources {
+	NPM1300_SOURCE_BUCK1,
+	NPM1300_SOURCE_BUCK2,
+	NPM1300_SOURCE_LDO1,
+	NPM1300_SOURCE_LDO2,
+};
+
+/* nPM1300 gpio control channels */
+enum npm1300_gpio_type {
+	NPM1300_GPIO_TYPE_ENABLE,
+	NPM1300_GPIO_TYPE_RETENTION,
+	NPM1300_GPIO_TYPE_PWM
+};
+
+/* nPM1300 regulator base addresses */
+#define BUCK_BASE 0x04U
+#define LDSW_BASE 0x08U
+
+/* nPM1300 regulator register offsets */
+#define BUCK_OFFSET_EN_SET    0x00U
+#define BUCK_OFFSET_EN_CLR    0x01U
+#define BUCK_OFFSET_PWM_SET   0x04U
+#define BUCK_OFFSET_PWM_CLR   0x05U
+#define BUCK_OFFSET_VOUT_NORM 0x08U
+#define BUCK_OFFSET_VOUT_RET  0x09U
+#define BUCK_OFFSET_EN_CTRL   0x0CU
+#define BUCK_OFFSET_VRET_CTRL 0x0DU
+#define BUCK_OFFSET_PWM_CTRL  0x0EU
+#define BUCK_OFFSET_SW_CTRL   0x0FU
+#define BUCK_OFFSET_VOUT_STAT 0x10U
+#define BUCK_OFFSET_CTRL0     0x15U
+
+/* nPM1300 ldsw register offsets */
+#define LDSW_OFFSET_EN_SET  0x00U
+#define LDSW_OFFSET_EN_CLR  0x01U
+#define LDSW_OFFSET_STATUS  0x04U
+#define LDSW_OFFSET_GPISEL  0x05U
+#define LDSW_OFFSET_CONFIG  0x07U
+#define LDSW_OFFSET_LDOSEL  0x08U
+#define LDSW_OFFSET_VOUTSEL 0x0CU
+
+struct regulator_npm1300_pconfig {
+	struct i2c_dt_spec i2c;
+	struct gpio_dt_spec dvs_state_pins[5];
+	uint8_t pad_val;
+};
+
+struct regulator_npm1300_config {
+	struct regulator_common_config common;
+	const struct device *p;
```

Comment:
Usually `const struct device *dev`

Reply:
This is a pointer to the parent device within a child device, so calling it dev may be confusing, e.g:
`[...] pconfig = dev->config->dev->config`

Maybe pdev?

Reply:
pdev sounds good.